File: Lab05Cpp/MonitoringSystem.cpp
#include "MonitoringSystem.hpp"
#include <chrono>
#include <iostream>

MonitoringSystem::MonitoringSystem() = default;

MonitoringSystem::MonitoringSystem(const MonitoringSystem& other)
    : m_rooms(other.m_rooms)
    , m_user(other.m_user)
    , m_logger(other.m_logger)
{
}

void MonitoringSystem::SetUser(const std::shared_ptr<User>& user)
{
    m_user = user;
}

void MonitoringSystem::AddRoom(const RoomPtr& room)
{
    m_rooms.push_back(room);
}

void MonitoringSystem::PollSensors()
{
    using clock = std::chrono::system_clock;
    for (const auto& room : m_rooms)
    {
        for (std::size_t i = 0; i < room->GetSensorCount(); ++i)
        {
            try
            {
                auto sensor = (*room)[i]; // тут может вылететь std::out_of_range
                float value = sensor->ReadValue();
                auto measurement = Measurement(sensor->GetId(), value, clock::now());

                m_logger += "Measurement: " + sensor->GetId();
                std::cout << measurement << "\n";

                // Если датчик - температура, проверим аномалию.
                auto tempPtr = std::dynamic_pointer_cast<TemperatureSensor>(sensor);
                if (tempPtr && tempPtr->IsAnomalous())
                {
                    Alert alert(
                        "A-" + sensor->GetId(),
                        sensor->GetId(),
                        "Temperature is out of range",
                        Severity::Critical
                    );

                    m_logger.Log("ALERT for sensor " + sensor->GetId());
                    if (m_user)
                    {
                        m_user->ReceiveAlert(alert);
                    }
                    alert.SendSoundNotification();
                }
            }
            catch (const std::exception& ex)
            {
                // Демонстрация обработки исключений: перехватываем любые std::exception
                // и логируем их.
                m_logger.Log(std::string("Ошибка при опросе датчика в комнате '") +
                             room->GetName() + "': " + ex.what());
            }
        }
    }
}

File: Lab04/Lab04/Lab04Monitoring/MonitoringSystem.cs
// Центральная система мониторинга.
// В этом классе показано:
//   * хранение комнат и пользователя;
//   * использование Logger;
//   * обработка исключений при опросе датчиков.
using System;
using System.Collections.Generic;

namespace Lab04Monitoring
{
    public class MonitoringSystem
    {
        private readonly List<Room> _rooms = new();
        private User? _user;

        public void SetUser(User user)
        {
            _user = user;
        }

        public void AddRoom(Room room)
        {
            _rooms.Add(room);
        }

        // Опрос всех датчиков с обработкой исключений.
        public void PollSensors()
        {
            foreach (var room in _rooms)
            {
                foreach (var sensor in room.Sensors)
                {
                    try
                    {
                        float value = sensor.ReadValue();
                        var measurement = new Measurement(sensor.Id, value, DateTime.Now);

                        Logger.Instance.Log($"Measurement from {sensor.Id} in room '{room.Name}' = {value:F2}");
                        Console.WriteLine(measurement);

                        if (sensor is TemperatureSensor temp && temp.IsAnomalous())
                        {
                            var alert = new Alert(
                                id: "A_" + sensor.Id,
                                sensorId: sensor.Id,
                                message: "Temperature is out of range",
                                severity: Severity.Critical);

                            Logger.Instance.Log($"ALERT for sensor {sensor.Id} in room '{room.Name}'");
                            _user?.ReceiveAlert(alert);
                            alert.SendSoundNotification();
                        }
                    }
                    catch (Exception ex)
                    {
                        // Демонстрация использования try/catch:
                        // здесь ловим любые ошибки, связанные с конкретным датчиком,
                        // и просто записываем их в лог, не падая всей программой.
                        Logger.Instance.Log($"Ошибка при опросе датчика {sensor.Id} в комнате '{room.Name}': {ex.Message}");
                    }
                }
            }
        }
    }
}
